LLM Benchmark Decoding Algorithm Rules Proposal.

yyetim opened the pull request.

szutenberg commented on inference_rules.adoc:
```diff
@@ -821,6 +821,30 @@ The DLRMv2 MLPerf inference code has an option to aggregate multiple consecutive
+=== LLM Benchmarks
+
+
+LLM benchmarks contain an auto-regressive generation loop. This loop’s execution is controlled by a decoding algorithm. The following rules discuss the decoding algorithm in the light of MLPerf Inference rules for the Closed submissions:
```
This new subsection is in FAQ section. Please keep Q&A format then.

szutenberg commented on inference_rules.adoc:
```diff
@@ -821,6 +821,30 @@ The DLRMv2 MLPerf inference code has an option to aggregate multiple consecutive
+
+LLM benchmarks contain an auto-regressive generation loop. This loop’s execution is controlled by a decoding algorithm. The following rules discuss the decoding algorithm in the light of MLPerf Inference rules for the Closed submissions:
+
+* Decoding algorithm should follow the implementation available on huggingface: https://huggingface.co/blog/how-to-generate#beam-search 
```
Please explain beam search algorithm in this document.

[https://huggingface.co/blog/how-to-generate#beam-search ](https://huggingface.co/blog/how-to-generate#beam-search) does not contain any implementation. Please describe how beam search works on a toy example and do not refer to HF docs. We believe that in the inference_rules the algorithm itself should be described, not its implementation.

In HF's beam_search algorithm we can find that topK with k=num_beams*2 is computed: [https://github.com/huggingface/transformers/blob/v4.30.0/src/transformers/generation/utils.py#L2955](https://github.com/huggingface/transformers/blob/v4.30.0/src/transformers/generation/utils.py#L2955). The document "How to generate beam search?" does not mention it and simplifies the algorithm a lot comparing with the actual implementation of beam_search in transformers library. How much freedom does submitter have?

szutenberg commented on inference_rules.adoc:
```diff
@@ -821,6 +821,30 @@ The DLRMv2 MLPerf inference code has an option to aggregate multiple consecutive
+LLM benchmarks contain an auto-regressive generation loop. This loop’s execution is controlled by a decoding algorithm. The following rules discuss the decoding algorithm in the light of MLPerf Inference rules for the Closed submissions:
+
+* Decoding algorithm should follow the implementation available on huggingface: https://huggingface.co/blog/how-to-generate#beam-search 
+  ** We are using beam=4 for this algorithm’s setting.
```
It's better to write that parameters of the decoding algorithm (in [https://github.com/mlcommons/inference/blob/master/language/gpt-j/backend.py#L13](https://github.com/mlcommons/inference/blob/master/language/gpt-j/backend.py#L13) ) must not be changed.

szutenberg commented on inference_rules.adoc:
```diff
@@ -821,6 +821,30 @@ The DLRMv2 MLPerf inference code has an option to aggregate multiple consecutive
+
+* Decoding algorithm should follow the implementation available on huggingface: https://huggingface.co/blog/how-to-generate#beam-search 
+  ** We are using beam=4 for this algorithm’s setting.
+  ** The output length determines the runtime of the generation. So, the algorithm in a submission should follow the same settings as the reference model and the algorithm should not change the output length.
```
Our proposal: remove “and the algorithm should not change the output length.” -> we just should follow the algorithm

szutenberg commented on inference_rules.adoc:
```diff
@@ -821,6 +821,30 @@ The DLRMv2 MLPerf inference code has an option to aggregate multiple consecutive
+* Decoding algorithm should follow the implementation available on huggingface: https://huggingface.co/blog/how-to-generate#beam-search 
+  ** We are using beam=4 for this algorithm’s setting.
+  ** The output length determines the runtime of the generation. So, the algorithm in a submission should follow the same settings as the reference model and the algorithm should not change the output length.
+* Summary of the allowed decoding algorithm:
```
No logical connection with the previous paragraphs, we believe that reference implementation already explains how does the full inference works so no point to duplicate it here.

szutenberg commented on inference_rules.adoc:
```diff
@@ -821,6 +821,30 @@ The DLRMv2 MLPerf inference code has an option to aggregate multiple consecutive
+  ** For the LLM benchmarks, each query contains a prompt. Upon receiving the query, the SUT runs this query in two stages: 
+    1. Run the forward pass of the model. As part of this, populate the KV-cache corresponding to this request.
+    2. Start the decoding loop using the beam search algorithm, using 4 beams.
+    3. Loop ends when the max tokens for the benchmark is reached or EOS token is detected as part of beam search.
```
Please describe early_stopping=True from HF::
Stop once:
•	Number of generated tokens in all beams reaches max_new_tokens=128 or
•	We found num_beams(=4) candidates.

szutenberg commented on inference_rules.adoc:
```diff
@@ -821,6 +821,30 @@ The DLRMv2 MLPerf inference code has an option to aggregate multiple consecutive
+  ** As explained earlier about caching, MLPerf Inference rules explicitly disallow caching of queries. KV-cache is a common technique used in caching the state of iterations of a single query in auto-regressive decoding. So, using a KV-cache, which doesn’t apply across queries, is allowed.
+  ** Entries of the KV-cache should be treated similar to activations of a forward pass. They can be quantized with respect to the quantization rules, and following the rules they can’t be pruned.
+  ** A KV-cache row is a tensor that deals with the execution of a single inference query. 
+  ** The entries corresponding to the query need to be reset and cannot retain state across new invocations.
```
Reset is very ambiguous term. Our proposal: KV-cache values must not be reused between any two inference calls (inference_call as in the reference implementation).

szutenberg commented on inference_rules.adoc:
```diff
@@ -821,6 +821,30 @@ The DLRMv2 MLPerf inference code has an option to aggregate multiple consecutive
+
+
+LLM benchmarks contain an auto-regressive generation loop. This loop’s execution is controlled by a decoding algorithm. The following rules discuss the decoding algorithm in the light of MLPerf Inference rules for the Closed submissions:
+
```
In the whole PR:
does "should" mean "should" from [https://www.ietf.org/rfc/rfc2119.txt](https://www.ietf.org/rfc/rfc2119.txt) or "must"?

szutenberg reviewed (changes requested):
Sorry, I won't be able to attend today meeting.

yyetim commented:
Thanks, I incorporated the changes from the LLM task force meeting, and also from the comments in the PR. PTAL.

We discussed in the task force meeting that Bhadri and Thomas were going to add more details to the beam search reference in the GPT-J. I added a TODO for them.

akhilarunkumar-dm commented on inference_rules.adoc:
```diff
@@ -821,6 +821,20 @@ The DLRMv2 MLPerf inference code has an option to aggregate multiple consecutive
+
+Q: MLPerf disallows caching queries. Is using a KV-cache in decoding allowed?
+
+A: Using a KV-cache is allowed in the same way as included in the reference model where it does not apply across queries. A KV-cache row is a tensor that must deal with the execution of a single inference query. Every input query must be computed in its entirety. Entries of the KV-cache must be treated similar to activations of a forward pass. They can be quantized with respect to the quantizatio ...
```
There was concern regarding the "cannot be pruned (or sparsified)" part of this in the inference working group today. Anton brought up a valid point that some implementations might choose to cache some entries while rematerializing some K/V vectors. I think that is a valid scenario that we want to allow. In my mind, we were not trying to prohibit that with the above rule.

yyetim replied:
Thanks for the note Akhil. Yes, I agree. We want to ensure with the rules that we standardize using the KV-cache as it has a lot of moving parts. However, not using a KV cache or rematerializing some of the values should be allowed. Pruning and sparsification changes model predictions and are disallowed in general as part of model equivalence rules. Not using KV-cache isn't pruning as long as the values are rematerialized.

nv-ananjappa commented:
@yyetim Could you sign the MLCommons CLA so that this PR can be merged? CC: @mrasquinha-g

yyetim commented:
Will do by tomorrow

On Mon, Jul 17, 2023, 11:53 PM Ashwin Nanjappa ***@***.***>
wrote:

> @yyetim <https://github.com/yyetim> Could you sign the MLCommons CLA so
> that this PR can be merged? CC: @mrasquinha-g
> <https://github.com/mrasquinha-g>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/mlcommons/inference_policies/pull/278#issuecomment-1638865671>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAGQZQU5WZTM2IC2KNFR7E3XQWQ3RANCNFSM6AAAAAAZZKINNQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

yyetim commented:
I submitted the form for the MLCommons CLA, @mrasquinha-g can you help with approval?

yyetim commented:
Requested access and added the google email used in the commit to my account. Will retry a dummy commit to rerun the check.

yyetim commented:
OK, the cla-bot check now passed. @mrasquinha-g can you approve again to merge? Thanks
